### DISCUSSION TITLE
Allow form field with required=False

### DISCUSSION THREAD

**Discussion by un1t** [django-recaptcha/django-recaptcha#264] [Q&A] [answered]
I use django-recaptcha==1.4.0
I tried to update to last version. But now required=True is hardcoded in form field code and validation code not check this parameter.
I need required=False, because I don't want to show captcha to all user, but only to specific users.

It is possible return this behaviour? Or maybe there is a simple workaround how to handle my case?

**Comment by AltusBarry** [comment, marked as answer]
Hello @un1t, we are sorry to hear you are having issues upgrading the django-recaptcha package.

Unfortunately, we do not foresee allowing a required toggle on the field.
Adding a recaptcha field means one wants to protect a specific aspect of your site against abuse.
Wanting to make it optional must be a deliberate implementation change, not a mere toggle.

The simplest way would be to just remove the captcha from the form's error list: 
```
class Form(forms.Form):
    captcha = RecaptchaField()

    def clean(self, *args, **kwargs):
        cleaned_data = super(Form, self).clean(*args, **kwargs)
        if not user.needs_captcha and "captcha" in self.errors.keys():
            return True
            self.errors.pop("captcha")
        return cleand_data
```

Other solutions include, but are not limited to; Not having the recaptcha field on the form by default and only adding it to the field list if the user needs it, subclassing the recaptcha field and overriding the __init__ behavior, altering the form's full_clean, altering the form's is_valid, etc.

**Comment by iogf** [comment]
I'm having issues to test a view because 

~~~python
    captcha = ReCaptchaField(required=False)
~~~
Doesn't work. I'm running automated tests and i would control required from a settings.py attribute
when in debug mode. My expected behavior of it would be that the captcha field doesn't
produce a form error when it is validated although it doesn't have a value in.

**Reply by HeartthrobRob** [reply]
I know this is late but Google provides test public / private key pairs that always approve, which can be used in offline environments. https://developers.google.com/recaptcha/docs/faq#id-like-to-run-automated-tests-with-recaptcha-v2-what-should-i-do
Since you already control the 'required' attribute in your settings.py, controlling these is not much additional effort. I'm using a context processor to send this detail to my templates.

**Comment by morenoh149** [comment]
I'm facing this as well with recaptcha v3. v3 doesn't provide an input for the user, instead I would like to do v3 plus server side validation. I'm seeing some bots automate the POST to register users.

you can just remove this line https://github.com/praekelt/django-recaptcha/blob/develop/captcha/fields.py#L47

**Comment by iogf** [comment]
Hi. I contoured the issue with https://github.com/iogf/shiva/blob/master/ticket_app/forms.py
The tests are now able to run.

**Comment by morenoh149** [comment]
I sidestepped the form field altogether as I'm comfortable on frontend https://github.com/praekelt/django-recaptcha/issues/217
